### ISSUE TITLE
OGC API Processes - Simple Profile

### ISSUE THREAD

**Issue by christophenoel** [opengeospatial/ogcapi-processes#171]
Hello,

Similarily to EOP for OGC O&M, or to  GML Simple Features, my intent would be to submit a branch with a Processes Simple Profile, which might be used as a reference for "simplified" implementation for prototypes, or any server not requiring all the legacy stuff.

One requirement for this Processes Simple Profile is that any request can be transparently submitted to a fully compliant implementation. Any response can be also understood by a fully compliant client implementation.

With this idea in mind, I would recommend a change in the **core specification** in execute.yaml for making optional some required fields:

```
required:
  - outputs
  - mode
  - response
```

I really would like to get rid of those prehistorical properties :) as mode (sync, async, auto) is an outdated approach (use callback instead), response (document, raw) is a useless option probably used by less than 1% of users, and outputs could be make optional as 99% of queries request all outputs.

As I wish to remove the mode and response fields in this Processes Simple Profile, this would preserve the compliance of the message with the core-spec.

Any people agreeing for this change ? :)

**Comment by bpross-52n** [comment]
@spacebel I invite you to the next SWG telecon on Monday to bring this up. Meeting info will be send over the SWG mailinglist.

**Comment by jerstlouis** [comment]
+1 for not makign required outputs, mode an response. This has also been discussed as a pre-requisiete for Routes harmonization.
When a single output is possible and it has a default format, requiring outputs is not necessary.
We also discussed mode and response preferences being specified by HTTP headers and having defaults instead.

**Comment by bpross-52n** [comment]
This is also linked to #127, which we originally postponed to version 1.1

**Comment by jerstlouis** [comment]
@bpross-52n @pvretano @christophenoel @cheazel
So I have thought a lot about this and here is what I think we should do and hereby propose:

**Considerations**
- Considering there is a desire for an OGC API-wide "async" vs. "sync" solution, that _Common_ could potentially define. This could also be used for example when a large coverage request takes a long time to complete, even without any special processing to do.
- Considering that with the _Workflows_ extension, the data requests are actually what may be triggering the processing
- Considering that regular _OGC API - Coverages, Features, Maps, Tiles, EDR_, etc. all return data synchronously by default
- Considering that we want to leverage [RFC 7240](https://tools.ietf.org/html/rfc7240) (client preference regarding sync vs. async)

**Proposal**
- If no `Prefer` header is set by the client, and no async callback mechanism is specified as part of a request, then the request SHALL be responded to synchronously
- If a server cannot or refuses to respond synchronously to a particular request (e.g. because it does not support synchronously responses, or a particular request would take too much resources or time), then it MAY respond with a 400 code. Some specifications could require that server implements synchronous support for reasonable requests.
- If a `Prefer` header is set, then the server SHOULD respond according to the client's hint
- This could be defined in a common manner across all APIs

*About output*
- `outputs` should be optional if the process is described as returning a single output which has only one or a default format

*About response*
- Suggesting that when responding synchronously, the raw response should always be what is returned. Processing servers doing a minimal amount of caching should be able to easily return right away the same exact response if the same request is given again. If the data is intended for another server (i.e. use case of passing the results around), then the server wishing to download the results can make the request itself directly. For multiple outputs, multipart responses can be used.
- When using async with polling, the job status already has a JSON structure. Could individual outputs be linked from an array of results directly in that JSON structure? Then that could point directly to the raw responses. Similar for other async mechanisms such as callback, it can simply point you to that job status which already includes the results directly linking to the raw responses. Not knowing whether (many servers we attempted to connect to so far do not follow the specifications right).

*Workflows response*
- For _Workflows_, we need two additional ways to (pre)-execute a process, expecting to get back either a collection description document or a landing page. A `response` query parameter would makse sense for that purpose, e.g. `response=collection` or `response=landingPage`. The actual "process this area/resolution of interest" request will be a follow-up `/coverage` or `/items` or `/tiles` or `/map` request (which itself could be sync or async based on the above proposal, i.e. omitting or including a `Prefer` header).

**Comment by fmigneault** [comment]
### Regarding the original propositions by @christophenoel 

I agree with parameters to be made optional with reasonable defaults, since most implementations tend to enforce only one way or another anyway until some need to support the other alternatives appear. I disagree though that they are useless or outdated. 
I feel that if some server does need to employs them, OGC should at least suggest *how* it should be done (expected names, formats, values, etc.) to have somewhat of a standard way to do it across instances. 

If find the values of the `mode` and `response` that I need to employ for CRIM's servers (`async`/`document`) are always the opposite of what @jerstlouis requires (`sync`/`raw`). This demonstrates that they are still relevant. 

I agree that `outputs` is not that useful for us as we also always want all the results. Users that don't want them can simply ignore the unnecessary ones, though I could understand how `raw` response could become quite massive if a lot of outputs are present. 
I don't have a strong opinion about that one.

### Regarding further propositions by @jerstlouis 

Just like execution endpoint was introduced to allow flexibility to any implementations that doesn't want to create a job for sync mode, I believe the option to allow implementations that want to always enforce async should also be respected. So I disagree with the proposal of : 

> If no Prefer header is set by the client, and no async callback mechanism is specified as part of a request, then the request SHALL be responded to synchronously (the server assumes that the client is not async-aware)

Also, async can well be valid when there is no callback mechanism in the request. The job submit response simply provides the link to where the job status can be located. 

The `Prefer` header is a good idea as it seems to be specifically designed for that kind of use case, and it also feels better than a query parameter. Still, it should remain a *preference*. The server should be the one to decide if it wants to respect that preference or outright ignore it. Now, should the server just ignore the preference or reply with something like `412 Precondition Failed` when it cannot fulfill the requested preference is another discussion. 

Regarding the direct `output`, I feel it adds yet again another ambiguity to what must be the expected output format. There is already many variants for an array/single value under an `outputs` field, then the same but with nested `value` fields before each item output, and another sometime with and sometimes without `value` depending of the literal/complex output type. Considering that it is quite common that the output ID is named `output`, it will become even more complicated to know if the body is being parsed correctly. I prefer to always extract the single output, but be sure I extracted it correctly from the body, than risk guessing incorrect parsing and have a long processing workflow fail because of inconsistent response formats.

**Comment by jerstlouis** [comment]
@fmigneault Note that this requirement to respond synchronously is paired with the permission that a server may respond with a 400 to such request if it wants to enforce async, so I think that is covered by the proposal.

I also agree with async being valid without callback mechanism -- specifying a response handler was one way for the client to tell the server it wants async, but using the `Prefer` header with `respond-async` is another way where a callback is not used.

About outputs, I believe we need to take a deeper look at all this and make sure it's crystal clear.
But in my opinion, an intermediate "result" document for a single output which is neither the job status page, nor the actual response, should not part of the solution, and definitely adds to the confusion.

**Comment by cristianoManuelLopes** [comment]
I wanted to avoid these discussions here in issues - we have SwG sessions for this - but since this is going on on multiple issues let me add my 0.02€...
For me WPS/API-Processes should be async by default, jobs should be created for every request, etc. sync calls should be an extension and should be made much simpler with GET (as per example the DAPA work in T16).

**Comment by jerstlouis** [comment]
@cristianoManuelLopes I disagree with async being the default because:
- It's much simpler to implement a sync-only client -- it can be as simple as a single cURL command.
- An OGC API-wide approach to sync/async needs to recognize the fact that all other OGC API so far default to sync.
- The use of the `Prefer` header allows the client to express awareness of async, and a preference for async or waiting x many seconds, but not to express that it does not support async. Therefore I suggest the omission of the header as indicating the client is not async-aware.
- The [harmonization](https://github.com/opengeospatial/ogcapi-routes/issues/15) with OGC API - Routes also expects sync to be the default

**Comment by fmigneault** [comment]
I believe also that default *should* be `async`, but I wouldn't mind to keep `auto` as default either, and let implementations pick whether `sync` or `async` should be their default in such case. 

I don't `async` feels like the right default only because it is how I'm using them, but also because it feels like the more reasonable and applicable way to do it considering how processing APIs are normally employed. There are not that many processes that only take a few seconds to complete. 

It is not reasonable in a real production environment to have a blocking call that basically locks the workers responding to requests until they complete the full process execution. With servers that are employed in high request demand, that means users would very often receive timeout or service unavailable responses just because all workers are busy replying to sync requests. I prefer the `async` submission approach, and workers will respond to the process execution queue in a timely manner. If fewer requests are made, the reply is almost immediate, and if not, you wait in whichever order the server deems fit. With `async`, the job is saved very quickly. This ensures workers can reply to very large amount of incoming processing requests.

Also, you don't necessarily need polling, a simple notification webhook can be employed. I don't believe the typical users that executes processes that take long minutes/hours in `sync` would just wait doing nothing until it completes. They send the request and go do something else in the meantime, and then come back once it is done. Having `sync` only forces them to keep the request session alive until it completes, which IMHO feels like a bad practice.

**Comment by pvretano** [comment]
FYI mode is gone.  There is no sync, async or auto anymore.  Instead the client and the server negotiate the execution mode using the Prefer header as per the RFC ... at least that is what the PR I am working on says right now.  I will be finishing it off this weekend as we will be reviewing in at Monday's SWG meeting.

**Comment by jerstlouis** [comment]
@fmigneault Some notes on your comments from my perspective...

> There are not that many processes that only take a few seconds to complete.

I am hoping to see processes in general complete order of magnitudes faster, e.g. with processing services staying "hot", low level native code, more optimizations, and leveraging GPUs. And I feel on-demand processing rather than batch-processing when possible might help make better use of computing resources, and ensure use of the latest available data. Many localized processing requests can easily complete under a second for a small area.

> It is not reasonable in a real production environment to have a blocking call that basically locks the workers responding to requests until they complete the full process execution.

The fact that a request is a sync HTTP request does not mean that it has to be a blocking call in the client. The client can asynchronously wait for the response.

> I don't believe the typical users that executes processes that take long minutes/hours in sync would just wait doing nothing until it completes.

I totally agree. But I believe in <1 second distributed processing calls, on the latest available data :)

**Comment by jerstlouis** [comment]
@pvretano If a client does not provide the `Prefer` header, what should it expect?
The `Prefer` header does not provide a mechanism to explicitly request synchronous (though one could possibly say `wait:` with a very large number). And how do we reconcile moving this up to _Common_, with _Coverages_ and _Features_ where the default behavior when no `Prefer` header is provided is a synchronous response.

We define a different default behavior per resource?

I thought omitting the `Prefer` header was a nice way to recognize a non-async aware client, but I might be the only one with this opinion... @cportele any opinion on this?

**Comment by cportele** [comment]
@jerstlouis 

My view is that it will often be the right approach for a resource to have a default behaviour (sync or async) and that would be specified in a req class (e.g., the Core req class of Features or Routes). Then there is another req class that adds support for the Prefer header and where the server will support both sync and async with the decision made by the server based on the input from the client (I think this is what Processes calls "auto").

Of course it is also ok to specify a resource (like Jobs) to require support for Prefer and sync/async execution in the base req class of the resource.

**Comment by jerstlouis** [comment]
What bothers me is that the use of `Prefer` implies a potentially async response, regardless of the preference provided.
Therefore if the default is not sync, there is no way for a client to ensure a synchronous response.

And I feel that although existing WPS implementations have traditionally been all async, as _OGC API - Processes_ go mainstream, tons of developers would want to quickly put together experimental clients and the simplicity of a single sync request makes things a lot easier and approachable.

**Comment by cristianoManuelLopes** [comment]
This discussion on sync and async on processing is very interesting one and for me relates to what I call "long-running" and "short-running" processing requests... for me it never is about making the API simpler, because for me API-Processes should not be a simple API since it requires job management and the like... so maybe yes we want a profile for the future that is sync focuses but for the core I would focus on the async case and consider job management and all of that as part of the spec!
Again this is my 0.02....

**Comment by christophenoel** [comment]
> There is no sync, async or auto anymore. Instead the client and the server negotiate the execution mode using the Prefer header

Very cool. I need to get up to date to this new feature.

Anyway, I'm very happy that the mentioned properties seem turning optional. This **simple profile** would help easing Techno Integration Experiments when the full API requires to effort to be implemented. It would typically only support the legacy "document" response, always "sync" and limit the I/O schemas to string and string arrays for example.

**Comment by fmigneault** [comment]
I would also like to have the *simple profile* item as future reference. Having people that exclusively focus and use the `sync` mode will help formulate simplified approaches and best pratices if we ever come around also supporting it on our side.

**Comment by pvretano** [comment]
@christophenoel the change to remove `mode` and use the `Prefer` header was rejected by the SWG because no consensus could be reached at this time.  So what I did in PR #182 was to make `mode` **optional**.  The default value of **mode** is the first item specified in the `jobControlOptions` array of the process description of each process.

**Comment by jerstlouis** [comment]
@pvretano @bpross-52n Regarding the `/results` discussion, that was also the specific topic of #178 .

I know we want to get this specification out the door as soon as possible, but the idea of the OGC API is that _Core_ should be as simple and straightforward to implement as possible, so having to later come up a profile that makes things simpler, or breaking things we realized were not really needed (e.g. getting rid of a separate `/results` as discussed in #178) after 1.0 is out, is not ideal in my opinion.

**Comment by jerstlouis** [comment]
See summary of current understanding of execute request possibilities, and the proposed simplification in 

https://github.com/opengeospatial/ogcapi-processes/issues/178#issuecomment-839447138

**Comment by jerstlouis** [comment]
Note that I believe this issue was filed before the feature freeze, and relates to the [Route API harmonization](https://github.com/opengeospatial/ogcapi-routes/issues/17), and for this we still need to consider:

- making `outputs` optional if a single output and it has a default response type
- making `response` optional. This is related to the clarification table as to whether it is needed at all -- the intent may have been to only have a meaning in sync mode, when I believe it should default to _raw_, while it should always be _document_ for async. Retrieving `/jobs/{jobId}/results` (not required in sync mode) should probably always return a JSON document (results.yaml) regardless of the execute request mode or response.

**Comment by bpross-52n** [comment]
Done, everything is now optional.
